### PR TITLE
[70_12] AST Syntax Highlighting for Scheme

### PR DESCRIPTION
**Pull request by UnbSky**
## What
AST Syntax Highlighting for Scheme.
6cdh's [tree-sitter-scheme](https://github.com/6cdh/tree-sitter-scheme) cannot distinguish between different symbols, which requires further optimization.
I fork 6cdh's [tree-sitter-scheme](https://github.com/6cdh/tree-sitter-scheme) and build a new  [tree-sitter-scheme](https://github.com/UnbSky/tree-sitter-scheme)

## Why
The old scheme highlighting does not consider partially multi-line code and cannot achieve colored brackets.
## How to test your changes?

1. Build and run.
2. Turn on"Edit" -> "Preference" -> "Other" -> "AST syntax highlighting".
3. New scheme code blocks and code lines will use AST syntax highlighting.

### CODE REVIEW

File: TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      "lazy-menu" "lazy-keyboard" "lazy-define" "lazy-define-macro" "lazy-initialize"  
+      "lazy-format" "lazy-language" "lazy-input-converter" "lazy-tmfs-handler"  
+      "$if")
+    (operator "{" "[" "(" ")" "]" "}" "'" ",")
+    (string_quote "\"")))
+
+(tm-define (parser-feature lan key)
+  (:require (and (== lan "scheme") (== key "brackets")))
+  `(,(string->symbol key)
+    (3 "(" ")")
```
Comment (da-liii) [inline TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm:111]: 这个3是什么意思？

Comment (da-liii) [reply TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm:111]: 看上去是三种嵌套层级，然后在下面定义了不同嵌套层级的颜色。这块需要简单写一下文档

Comment (da-liii) [reply TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm:111]: 看上去是三种嵌套层级，然后在下面定义了不同嵌套层级的颜色。这块需要简单写一下文档

File: src/System/Language/ast_language.cpp [hunk 2]
```diff
@@ -104,6 +107,22 @@ ast_language_rep::customize_highlight_theme (tree config) {
   }
 }
 
+void
+ast_language_rep::customize_special_symbol (tree config) {
+  lang_ast_parser->reset_intercept_symbols ();
+  tokenize_set= hashset<string> ();
+  for (int i= 0; i < N (config); i++) {
```
Comment (da-liii) [inline src/System/Language/ast_language.cpp:114]: 可以直接写 for (tree group_of : config)

File: src/System/Language/ast_language.cpp [hunk 3]
```diff
@@ -104,6 +107,22 @@ ast_language_rep::customize_highlight_theme (tree config) {
...
+    tree   group_of    = config[i];
+    string special_type= get_label (group_of);
+    for (int j= 0; j < N (group_of); j++) {
```
Comment (da-liii) [inline src/System/Language/ast_language.cpp:117]: 可以用 range-for

File: TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+;; COPYRIGHT   : (C) 2024  UnbSky
+;;
+;; This software falls under the GNU general public license version 3 or later.
+;; It comes WITHOUT ANY WARRANTY WHATSOEVER. For details, see the file LICENSE
+;; in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(texmacs-module (code scheme-ast-lang)
+  (:use (prog default-lang)))
```
Comment (da-liii) [inline TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm:14]: 这个use应该不需要吧，default-lang提供的是老的parser-feature的默认值

File: TeXmacs/plugins/code/progs/code/cpp-ast-lang.scm [hunk 5]
```diff
@@ -59,6 +59,11 @@
     (3 "(" ")")
     (3 "[" "]")
     (3 "{" "}")))
+
+(tm-define (parser-feature lan key)
```
Comment (da-liii) [inline TeXmacs/plugins/code/progs/code/cpp-ast-lang.scm:63]: 这个看上去是空的，应该放在 default-ast-lang.scm 里面，然后最前面的 `(:use (prog default-lang))`应该改为 `(:use (prog default-ast-lang))`

File: src/System/Language/ast_language.cpp (hunk 3 above)
Comment (UnbSky) [reply src/System/Language/ast_language.cpp:117]: 已修改该部分为 range-for

File: TeXmacs/plugins/code/progs/code/cpp-ast-lang.scm (hunk 5 above)
Comment (UnbSky) [reply TeXmacs/plugins/code/progs/code/cpp-ast-lang.scm:63]: 已添加default-ast-lang.scm

File: TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm (hunk 1 above)
Comment (UnbSky) [reply TeXmacs/plugins/code/progs/code/scheme-ast-lang.scm:111]: 已将该部分移入default-ast-lang.scm，因为目前cpp和scheme的彩色括号周期和匹配括号相同，该部分已添加注释说明

**Comment by da-liii** [comment]
LGTM, and only your mentor can approve your pull request

**Comment by da-liii** [comment]
合并一下 branch-1.2 分支，ci就可以通过了

**Pull request merged**